agent/consul/state/catalog.go

```diff
@@ -4628,6 +4690,61 @@ func (s *Store) combinedServiceNodesTxn(tx ReadTxn, ws memdb.WatchSet, names []s
 	return maxIdx, resp, nil
 }
 
+func upstreamsFromGWTxn(tx ReadTxn, service structs.ServiceName) ([]structs.ServiceName, error) {
```

Comment:
Forgot to actually add this comment to the last review.

Last change that I see, but can you try and standardize the naming convention for the two new helper funcs? Since the other is named:

`downstreamGatewaysForServiceTxn`

maybe:

`upstreamServicesForGatewayTxn`?